File: acme-java-reactive/src/main/java/acme/controller/ApiController.java
package acme.controller;

import acme.dto.AccountDto;
import acme.dto.CustomerDto;
import acme.dto.TransactionRequest;
import acme.dto.TransactionResponse;
import acme.entity.Transaction;
import acme.service.CustomerService;
import acme.service.TransactionService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api")
public class ApiController {

    private final CustomerService customerService;
    private final TransactionService transactionService;

    public ApiController(CustomerService customerService, TransactionService transactionService) {
        this.customerService = customerService;
        this.transactionService = transactionService;
    }

    @GetMapping("/customers/{customerId}")
    public Mono<ResponseEntity<CustomerDto>> getCurrentCustomer(@PathVariable UUID customerId) {
        return customerService.getCurrentCustomer(customerId)
            .map(ResponseEntity::ok)
            .defaultIfEmpty(ResponseEntity.notFound().build());
    }

    @GetMapping("/customers/{customerId}/accounts")
    public Mono<ResponseEntity<Flux<AccountDto>>> getCustomerAccounts(@PathVariable UUID customerId) {
        return Mono.just(ResponseEntity.ok(customerService.getCustomerAccounts(customerId)));
    }

    @PostMapping("/transactions")
    public Mono<ResponseEntity<TransactionResponse>> performTransaction(@RequestBody TransactionRequest request) {
        return transactionService.performTransaction(request)
            .map(ResponseEntity::ok);
    }

    @GetMapping("/accounts/{accountId}/transactions")
    public Mono<ResponseEntity<Flux<Transaction>>> getTransactionHistory(@PathVariable UUID accountId) {
        return Mono.just(ResponseEntity.ok(transactionService.getTransactionHistory(accountId)));
    }

    @GetMapping("/accounts/total-amount")
    public Mono<ResponseEntity<Map<String, Long>>> getTotalAccountsAmount() {
        return customerService.getTotalAccountsAmount()
            .map(totalAmount -> ResponseEntity.ok(Map.of("totalAmount", totalAmount)));
    }
}

File: acme-springboot/src/main/kotlin/acme/controller/ApiController.kt
package acme.controller

import acme.dto.AccountDto
import acme.dto.CustomerDto
import acme.dto.TransactionRequest
import acme.entity.Transaction

import acme.service.CustomerService
import acme.service.TransactionService
import io.swagger.v3.oas.annotations.OpenAPIDefinition
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api")
class ApiController(
    private val customerService: CustomerService,
    private val transactionService: TransactionService
) {

    private val logger = LoggerFactory.getLogger(this::class.java)

    @GetMapping("/customers/{customerId}")
    fun getCurrentCustomer(@PathVariable customerId: UUID): ResponseEntity<CustomerDto> {
        val customer = customerService.getCurrentCustomer(customerId)
        return if (customer != null) {
            ResponseEntity.ok(customer)
        } else {
            ResponseEntity.notFound().build()
        }
    }

    @GetMapping("/customers/{customerId}/accounts")
    fun getCustomerAccounts(@PathVariable customerId: UUID): ResponseEntity<List<AccountDto>> {
        return try {
            val accounts = customerService.getCustomerAccounts(customerId)
            ResponseEntity.ok(accounts)
        } catch (_: IllegalArgumentException) {
            ResponseEntity.notFound().build()
        }
    }

    @PostMapping("/transactions")
    fun performTransaction(@RequestBody request: TransactionRequest): ResponseEntity<Any> {
        return try {
            val response = transactionService.performTransaction(request)
            ResponseEntity.ok(response)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to (e.message ?: "Invalid request")))
        } catch (e: Exception) {
            logger.error("Transaction failed", e)
            ResponseEntity.internalServerError().body(mapOf("error" to "Transaction failed"))
        }
    }

    @GetMapping("/accounts/{accountId}/transactions")
    fun getTransactionHistory(@PathVariable accountId: UUID): ResponseEntity<List<Transaction>> {
        val transactions = transactionService.getTransactionHistory(accountId)
        return ResponseEntity.ok(transactions)
    }

    @GetMapping("/accounts/total-amount")
    fun getTotalAccountsAmount(): ResponseEntity<Map<String, Long>> {
        val totalAmount = customerService.getTotalAccountsAmount()
        return ResponseEntity.ok(mapOf("totalAmount" to totalAmount))
    }
}
